src/Microsoft.ML.Core/Data/ColumnTypeExtensions.cs

```diff
@@ -22,5 +22,48 @@ internal static class ColumnTypeExtensions
         /// Zero return means either it's not a key type or the cardinality is unknown.
         /// </summary>
         public static int GetKeyCount(this ColumnType columnType) => (columnType as KeyType)?.Count ?? 0;
+
+        /// <summary>
+        /// For non-vector types, this returns the column type itself (i.e., return <paramref name="columnType"/>).
+        /// For vector types, this returns the type of the items stored as values in vector.
+        /// </summary>
+        public static ColumnType GetItemType(this ColumnType columnType) => (columnType as VectorType)?.ItemType ?? columnType;
+
+        /// <summary>
+        /// Zero return means either it's not a vector or the size is unknown.
+        /// </summary>
+        public static int GetVectorSize(this ColumnType columnType) => (columnType as VectorType)?.Size ?? 0;
```

Comment:
Just something I found interesting. When I saw this code, I thought to myself that maybe something like an `is` would actually work faster. That is, instead of `(columnType as VectorType)?.Size ?? 0` prefer `columnType is VectorType vecType ? vecType.Size : 0`. But, I benchmarked an analogous situation, and I cannot see that one is faster than the other. So, this code is fine, but I found the result surprising. 😄

Reply:
I guess I didn't dig into the IL generated before - but my assumption was that the IL would be pretty similar.

My benchmark showed they are basically identical:

```C#
    public class TypeChecking
    {
        private static ColumnType t1 = BoolType.Instance;
        private static ColumnType t2 = NumberType.R4;
        private static ColumnType t3 = new VectorType(BoolType.Instance, 5);

        [Benchmark]
        public int AsTrue()
        {
            return (t3 as VectorType)?.Size ?? 0;
        }

        [Benchmark]
        public int IsTrue()
        {
            return t3 is VectorType vectorType ? vectorType.Size : 0;
        }

        [Benchmark]
        public int AsFalse()
        {
            return (t2 as VectorType)?.Size ?? 0;
        }

        [Benchmark]
        public int IsFalse()
        {
            return t2 is VectorType vectorType ? vectorType.Size : 0;
        }
    }
```

```
 Method |     Mean |     Error |    StdDev |
-------- |---------:|----------:|----------:|
  AsTrue | 1.701 ns | 0.0603 ns | 0.0564 ns |
  IsTrue | 1.703 ns | 0.0359 ns | 0.0336 ns |
 AsFalse | 1.248 ns | 0.0271 ns | 0.0254 ns |
 IsFalse | 1.245 ns | 0.0386 ns | 0.0342 ns |
```

And here's the generated IL:

```IL
    .method public hidebysig instance int32 AsTrue () cil managed 
    {
        .custom instance void [BenchmarkDotNet]BenchmarkDotNet.Attributes.BenchmarkAttribute::.ctor() = (
            01 00 00 00
        )
        IL_0000: ldsfld class [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.ColumnType RandomBenchmarks.TypeChecking::t3
        IL_0005: isinst [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.VectorType
        IL_000a: dup
        IL_000b: brtrue.s IL_0010

        IL_000d: pop
        IL_000e: ldc.i4.0
        IL_000f: ret

        IL_0010: call instance int32 [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.VectorType::get_Size()
        IL_0015: ret
    }

   .method public hidebysig instance int32 IsTrue () cil managed 
    {
        .custom instance void [BenchmarkDotNet]BenchmarkDotNet.Attributes.BenchmarkAttribute::.ctor() = (
            01 00 00 00
        )
        .locals init (
            [0] class [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.VectorType V_0
        )

        IL_0000: ldsfld class [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.ColumnType RandomBenchmarks.TypeChecking::t3
        IL_0005: isinst [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.VectorType
        IL_000a: dup
        IL_000b: stloc.0
        IL_000c: brtrue.s IL_0010

        IL_000e: ldc.i4.0
        IL_000f: ret

        IL_0010: ldloc.0
        IL_0011: callvirt instance int32 [Microsoft.ML.Core]Microsoft.ML.Runtime.Data.VectorType::get_Size()
        IL_0016: ret
    }
```

Reply:
Yup, I observed the same thing. The IL for `foo?.Bar ?? biz` is a lot smarter than I thought it might be.